### PR TITLE
Sub: send AP_ADC

### PR DESCRIPTION
**Pull request by jaxxzer**
This required some tweaks to the channel() api on PX4.

In master there is currently no way for global access to adc values in HAL api as far as I could find.

This will work for #877 on Sub.

### CODE REVIEW

**Comment by jaxxzer** [comment]
Agreed on the need for a new interface.

File: ArduSub/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -256,6 +257,28 @@ NOINLINE void Sub::send_extended_status1(mavlink_channel_t chan)
...
 
+    // FMUV2 support only
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4
+    static const uint8_t max_channels = 6;
+    static const uint8_t pins[max_channels] = { 15, 14, 13, 12, 11, 10 };
+
+    float adcs[max_channels];
+
+    for (uint8_t i = 0; i < max_channels; i++) {
+        AP_HAL::AnalogSource* adc = hal.analogin->channel(pins[i]);
```
[the remark this answers is not in the document]
Comment (jaxxzer) [reply ArduSub/GCS_Mavlink.cpp]: The memory is only allocated once, I think.

File: libraries/AP_HAL_PX4/AnalogIn.cpp [hunk 2]
```diff
@@ -380,14 +380,16 @@ void PX4AnalogIn::_timer_tick(void)
 
 AP_HAL::AnalogSource* PX4AnalogIn::channel(int16_t pin) 
 {
-    for (uint8_t j=0; j<PX4_ANALOG_MAX_CHANNELS; j++) {
+    uint8_t j;
+    for (j=0; j<PX4_ANALOG_MAX_CHANNELS; j++) {
         if (_channels[j] == nullptr) {
```
Comment (jaxxzer) [inline libraries/AP_HAL_PX4/AnalogIn.cpp]: See here.

Comment (tridge) [reply libraries/AP_HAL_PX4/AnalogIn.cpp]: that is not indexed by pin. I don't see what stops it allocating every time

Comment (jaxxzer) [reply libraries/AP_HAL_PX4/AnalogIn.cpp]: It only allocates if the item at the current index is null. There are only PX4_ANALOG_MAX_CHANNELS elements, this is the maximum AnalogSources that will be allocated.

Comment (tridge) [reply libraries/AP_HAL_PX4/AnalogIn.cpp]: ok, and when that fails then the adc send function in sub will crash with a null ptr dereference

Comment (jaxxzer) [reply libraries/AP_HAL_PX4/AnalogIn.cpp]: It never returns null.

**Comment by jaxxzer** [comment]
I've fixed the potential for a buffer overflow. I've also added a guard in Sub to make sure there never will be any segfault in the case that someone does something silly in the future.

There never was any memory leak or null ptr dereferencing with the original patch.

File: ArduSub/GCS_Mavlink.cpp [hunk 3]
```diff
@@ -256,6 +257,32 @@ NOINLINE void Sub::send_extended_status1(mavlink_channel_t chan)
         0, // comm drops in pkts,
         0, 0, 0, 0);
 
+    // FMUV2 support only
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4
```
Comment (OXINARF) [inline ArduSub/GCS_Mavlink.cpp]: This only says it for the PX4 HAL, doesn't say anything about the board . and with hard-coded pin numbers you definitely want to check for the correct board. I guess that adding another macro to check if CONFIG_ARCH_BOARD_PX4FMU_V2 is defined is enough.

**Comment by jaxxzer** [comment]
This is put on hold until we have a new ADC interface in HAL: https://github.com/ArduPilot/ardupilot/issues/8527

**Pull request closed without merging**